src/fsharp/FSharp.Core/prim-types.fs

```diff
@@ -1871,7 +1871,12 @@ namespace Microsoft.FSharp.Core
             module mos =
                 type IGetType =
```

Comment:
Looking through all this code I realize now that #513 really was not properly code reviewed at all.  GitHub simply didn't show the diff for the critical file prim-types.fs and now comments or proper review has been done.  This is a grave problem with relying on GitHub PRs for code review.

For example two basic things that indicate this code hasn't been reviewed:

* We need proper /// comments on every type declaration in prim-types

* Please don't use abc_def_hij style naming 

I'm not sure what to do about this yet.  We should really revert the commit and start again with a full code review, but I think that @KevinRansom didn't want to do that.  I do understand that the testing is good and the performance results are real.  But we need to be more considered here.